roles/translator/src/lib/mod.rs

```diff
@@ -1,7 +1,204 @@
+            {
+                Ok(_) => info!("Connected to Upstream!"),
+                Err(e) => {
+                    error!("Failed to connect to Upstream EXITING! : {}", e);
+                panic!();
+                }
+            }
+
+            // Start receiving messages from the SV2 Upstream role
+            if let Err(e) = upstream_sv2::Upstream::parse_incoming(upstream.clone()) {
+                error!("failed to create sv2 parser: {}", e);
+                panic!();
+            }
+
+            debug!("Finished starting upstream listener");
+            // Start task handler to receive submits from the SV1 Downstream role once it connects
+            if let Err(e) = upstream_sv2::Upstream::handle_submit(upstream.clone()) {
+                error!("Failed to create submit handler: {}", e);
+                return;
+            }
+
+            // Receive the extranonce information from the Upstream role to send to the Downstream role
+            // once it connects also used to initialize the bridge
+            let (extended_extranonce, up_id) = rx_sv2_extranonce.recv().await.unwrap();
+            loop {
+                let target: [u8; 32] = target.safe_lock(|t| t.clone()).unwrap().try_into().unwrap();
+                if target != [0; 32] {
+                    break;
+                };
+                async_std::task::sleep(std::time::Duration::from_millis(100)).await;
+            }
+
+            // Instantiate a new `Bridge` and begins handling incoming messages
+            let b = proxy::Bridge::new(
+                rx_sv1_downstream,
+                tx_sv2_submit_shares_ext,
+                rx_sv2_set_new_prev_hash,
+                rx_sv2_new_ext_mining_job,
+                tx_sv1_notify.clone(),
+                status::Sender::Bridge(tx_status.clone()),
+                extended_extranonce,
+                target,
+                up_id,
+            );
+            proxy::Bridge::start(b.clone());
+
+            // Format `Downstream` connection address
+            let downstream_addr = SocketAddr::new(
+                IpAddr::from_str(&config.downstream_address).unwrap(),
+                config.downstream_port,
+            );
+
+            // Accept connections from one or more SV1 Downstream roles (SV1 Mining Devices)
+            downstream_sv1::Downstream::accept_connections(
+                downstream_addr,
+                tx_sv1_bridge,
+                tx_sv1_notify,
+                status::Sender::DownstreamListener(tx_status.clone()),
+                b,
+                downstream_difficulty_config,
+                diff_config,
+            );
+        });
+
+    // Check all tasks if is_finished() is true, if so exit
+    loop {
+        let task_status: status::Status = rx_status.recv().fuse().await.unwrap();
+
+        match task_status.state {
+            // Should only be sent by the downstream listener
+            status::State::DownstreamShutdown(err) => {
```

Comment:
Does this mean other roles determine when the Translator should shutdown?

Reply:
https://github.com/stratum-mining/stratum/pull/1066#discussion_r1729539914